### PR TITLE
GH-35926: [C++][Parquet] Allow disabling ColumnIndex by disabling statistics

### LINKED ISSUE #35926: [C++][Parquet] Parquet PageIndex allow only building Offset Index
### Describe the enhancement requested

Currently, if `page_index` is enabled, the Offset-index and Column-index would be built.

However, sometimes, user only need offset-index to find the pages, and statistics is not useful, we're sure these column has bad distribution and better not collect there stats. So, here we can allow only build offset index.

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
When statistics are disabled for a column, then only OffsetIndex should be emitted, not ColumnIndex.
Also, fix a bug in the `ColumnProperties` constructor where `page_index_enabled` would not be forwarded correctly.

* Closes: #35926

### CODE REVIEW

**Comment by mapleFU** [comment]
cc @wgtmac @pitrou

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 1]
```diff
@@ -5350,6 +5352,30 @@ TEST_F(ParquetPageIndexRoundTripTest, SimpleRoundTrip) {
                             /*null_counts=*/{1}}));
 }
 
+TEST_F(ParquetPageIndexRoundTripTest, SimpleRoundTripWithStatsDisabled) {
+  auto writer_properties = WriterProperties::Builder()
+                               .enable_write_page_index()
+                               ->disable_statistics()
```
Comment (wgtmac) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5358]: Should we test this per column?

Comment (mapleFU) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5359]: Added now

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 2]
```diff
@@ -5370,12 +5370,53 @@ TEST_F(ParquetPageIndexRoundTripTest, SimpleRoundTripWithStatsDisabled) {
...
+    ])"}));
+
+  ReadPageIndexes(/*expect_num_row_groups=*/2, /*expect_num_pages=*/1);
+
+  ColumnIndexObject empty_column_index{};
+  empty_column_index.empty = true;
+  EXPECT_THAT(
+      column_indexes_,
+      ::testing::ElementsAre(
+          empty_column_index,
```
Comment (wgtmac) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5404]: Why add a new `empty` flag? Isn't `ColumnIndexObject{}` enough to represent a missing ColumnIndex?

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 3]
```diff
@@ -5356,7 +5357,6 @@ TEST_F(ParquetPageIndexRoundTripTest, SimpleRoundTripWithStatsDisabled) {
   auto writer_properties = WriterProperties::Builder()
                                .enable_write_page_index()
                                ->disable_statistics()
-                               ->max_row_group_length(4)
```
[the remark this answers is not in the document]
Comment (mapleFU) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: No need to split rowgroup here, just ignore it

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 2 above)
Comment (mapleFU) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5404]: Just want to make clear that it's explicit built from a `nullptr`

Comment (wgtmac) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5404]: That seems duplicate. A valid ColumnIndex must have at least one page. So `null_pages`, `min_values` and `max_values` will not be empty and share the same size if ColumnIndexObject is created from a non null ColumnIndex.

Comment (mapleFU) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5404]: Updated

File: cpp/src/parquet/properties.h [hunk 4]
```diff
@@ -154,7 +154,7 @@ class PARQUET_EXPORT ColumnProperties {
         statistics_enabled_(statistics_enabled),
         max_stats_size_(max_stats_size),
         compression_level_(Codec::UseDefaultCompressionLevel()),
-        page_index_enabled_(DEFAULT_IS_PAGE_INDEX_ENABLED) {}
+        page_index_enabled_(page_index_enabled) {}
```
Comment (wgtmac) [inline cpp/src/parquet/properties.h:157]: Good catch!

**Review by wgtmac — APPROVED** [review]
LGTM

**Comment by mapleFU** [comment]
@pitrou Mind take a look?

**Comment by mapleFU** [comment]
@pitrou @wgtmac Do you still have any comments?

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 5]
```diff
@@ -5350,6 +5350,70 @@ TEST_F(ParquetPageIndexRoundTripTest, SimpleRoundTrip) {
...
+      [2,     "b",  [1, 2]   ],
+      [3,     "c",  [null]   ],
+      [null,  "d",  []       ],
+      [5,     null, [3, 3, 3]],
+      [6,     "f",  null     ]
+    ])"}));
+
+  ReadPageIndexes(/*expect_num_row_groups=*/1, /*expect_num_pages=*/1);
+  for (auto& column_index : column_indexes_) {
+    // Means page is empty.
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5372]: Which page is empty? I don't understand this comment.

Comment (pitrou) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:5372]: Why not test against `ColumnIndexObject{}` instead?

Comment (mapleFU) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: Sure, changed

**Review by pitrou — APPROVED** [review]
+1, thanks @mapleFU !

**Comment by pitrou** [comment]
The remaining CI failures are unrelated, will merge.

**Pull request merged**